### PR TITLE
Update to use new Marain tenancy approach.

### PR DESCRIPTION
**Pull request by jongeorge1**
This update allows Marain.Operations to use the new approach to tenancy in Marain.

As part of the PR I've added a couple of new projects which don't belong here; my intention is to get to the point where everyone is happy with the approach and then move them to where they belong prior to final approval.

There are two parts to the PR:
1. The operations services need to validate that the incoming tenant represents a client that is enrolled to use them.
2. Tests need to be able to create tenants on the fly and clean them up once the test is done.

To address the first part, I have added the project `Marain.Services.Tenancy`. This will be moved to `Marain.TenantManagement` once it's ready. It provides access to standard operations that Marain services will need to use to use incoming tenant information.

Firstly, `MarainServiceConfiguration` holds the tenant Id of the service, which is intended to be read from config (the `IServiceCollection.AddMarainServiceConfiguration` method encapsulates this. It's here because a Marain service will need to know its own Service Tenant Id in order to validate enrollment and obtain delegated tenants. It's unlikely a service will need to use this directly once it's set up.

Then `IMarainServiceTenancyHelper` provides a facade over `ITenantManagementService`, exposing just the features needed by services.

The `IMarainServiceTenancyHelper` will be consumed by the services themselves (i.e. the classes implementing `IOpenApiService`, and it's expected that the first thing the majority of operations do will be to validate the incoming tenant Id via `GetRequestingTenantAsync`.

Ideally I would have done something that could have been plugged into the Menes pipeline, but there isn't a convenient point to do this at the moment - context builder components execute a bit too early in the pipeline to be convenient. I think down the line we should look at improving this so that tenant identification and validation can be added globally.

Because we're expecting this class to be consumed by Marain services, the `GetRequestingTenantAsync` method throws exceptions from the `Menes.Exceptions` namespace directly.

To address the second part I've added the project `Marain.TenantManagement.Testing`. This contains the `InMemoryTenantProvider` to make it easier for self-contained specs that run in-process (i.e. they don't run a function and they don't try to talk to other Marain functions that are running in the cloud). It also contains `TransientTenantManager`, a class which is intended to help create transient service and client tenants, and clean them up once the test is complete.

### CODE REVIEW

File: Solutions/Marain.Services.Tenancy/Marain/Services/Tenancy/IMarainServiceTenancyHelper.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{
+    using System.Threading.Tasks;
+    using Corvus.Tenancy;
+    using Corvus.Tenancy.Exceptions;
+    using Menes.Exceptions;
+
+    /// <summary>
+    /// Provides methods required by Marain services to validate and work with tenants.
+    /// </summary>
+    public interface IMarainServiceTenancyHelper
```
Comment (mwadams) [inline Solutions/Marain.Services.Tenancy/Marain/Services/Tenancy/IMarainServiceTenancyHelper.cs:15]: I have to say I don't like `Helper`

Isn't this just `IMarainServiceTenancyManagement` (to reflect `ITenantManagement`)

File: Solutions/Marain.Services.Tenancy/Marain/Services/Tenancy/Internal/MarainServiceTenancyHelper.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            // Validate it's of the expected type. This will throw an ArgumentException if the tenant is not of the expected
+            // type. This is not particularly useful, so we will catch this and instead throw an exception that will result
+            // in a Bad Request response.
+            try
+            {
+                tenant.EnsureTenantIsOfType(MarainTenantType.Client, MarainTenantType.Delegated);
+            }
+            catch (ArgumentException)
+            {
+                throw new OpenApiNotFoundException($"The specified tenant Id, '{tenantId}', is of the wrong type for this request");
```
Comment (mwadams) [inline Solutions/Marain.Services.Tenancy/Marain/Services/Tenancy/Internal/MarainServiceTenancyHelper.cs:49]: Is this really the right exception to throw? You say you want "Bad Request" but surely that will produce a "Not Found"?

Comment (jongeorge1) [reply Solutions/Marain.Services.Tenancy/Marain/Services/Tenancy/Internal/MarainServiceTenancyHelper.cs:49]: Yes. Originally I was going to return BadRequest, but @idg10 and I talked this through and there are a couple of reasons for the resulting change.

Firstly, the existence of anything other than client tenants is an internal implementation detail of the platform. So saying "we found a tenant with the specified ID, but it's not valid for this scenario" is effectively leaking some of our implementation detail.

Following on from this, the path for these requests is generally of the form `/{client tenant Id}/marain/something/whatever`. If your Id doesn't match a client tenant then the response is "there isn't a client tenant with that Id", which better maps onto a 404.

Comment (jongeorge1) [reply Solutions/Marain.Services.Tenancy/Marain/Services/Tenancy/Internal/MarainServiceTenancyHelper.cs:49]: Oh. The comment is wrong.

**Comment by jongeorge1** [comment]
Changing this PR to ready for review as I've now pulled out the two new projects and moved them into the `Marain.TenantManagement` repo.

**Pull request merged**